src/odb/src/lefout/lefout.cpp

```diff
@@ -135,32 +137,104 @@ void lefout::writeHeader(dbBlock* db_block)
   writeUnits(/*database_units = */db_block->getDbUnitsPerMicron());
 }
 
-void lefout::writeObstructions(dbBlock* db_block)
+void lefout::writeObstructions(dbBlock* db_block, int bloat)
 {
-  std::set<dbTechLayer*> obstruction_layers;
-  getTechLayerObstructions(db_block, obstruction_layers);
+  std::map<dbTechLayer*,std::set<dbObstruction*>> obstructions;
+  getObstructions(db_block, obstructions);
+  mergeObstructions(db_block, bloat, obstructions);
 
   fprintf(_out, "  OBS\n");
-  dbBox* block_bounding_box = db_block->getBBox();
-  for (dbTechLayer* tech_layer : obstruction_layers) {
+
+  for (auto const& pair : obstructions) {
+    dbTechLayer* tech_layer = pair.first;
+    std::set<dbObstruction*> obstructions_per_layer = pair.second;
+
     fprintf(_out, "    LAYER %s ;\n", tech_layer->getName().c_str());
-    writeBox("  ", block_bounding_box);
+    for (dbObstruction* obs : obstructions_per_layer) {
+      writeBox("  ", obs->getBBox());
+    }
   }
+  
   fprintf(_out, "  END\n");
 }
 
-void lefout::getTechLayerObstructions(
+void lefout::mergeObstructions(
+  dbBlock* block,
+  int bloat,
+  std::map<dbTechLayer*,std::set<dbObstruction*>>& obstructions) const
+{
+  using namespace boost::polygon::operators;
+
+  for ( const auto &pair : obstructions ) {
+    dbTechLayer* tech_layer = pair.first;
+    std::set<dbObstruction*> obstructions_per_layer = pair.second;
+
+    int pitch = tech_layer->getPitch();
+
+    // Compose all the rects in the layer to merged polygon with bloat
+    boost::polygon::polygon_90_set_data<int> polySet;
+    for (dbObstruction* obs : obstructions_per_layer) {
+      Rect rect = obs->getBBox()->getBox();
+      boost::polygon::polygon_90_set_data<int> poly;
+      poly = boost::polygon::rectangle_data<int>{rect.xMin(), rect.yMin(), rect.xMax(), rect.yMax()};
+      polySet += poly.bloat(bloat*pitch,bloat*pitch,bloat*pitch,bloat*pitch);
+    }
+    polySet.shrink2(bloat*pitch,bloat*pitch,bloat*pitch,bloat*pitch);
+  
+    // Decompose the polygon set to rectanges in non-preferred direction
+    std::vector<boost::polygon::rectangle_data<int>> rects;
+    if (strcasecmp(tech_layer->getDirection().getString(),"HORIZONTAL")) {
+      polySet.get_rectangles(rects, boost::polygon::VERTICAL); 
+    } else if (strcasecmp(tech_layer->getDirection().getString(),"VERTICAL")) {
+      polySet.get_rectangles(rects, boost::polygon::HORIZONTAL); 
+    } else if (strcasecmp(tech_layer->getDirection().getString(),"NONE")) {
+      polySet.get_rectangles(rects); 
+    }
+
+    // Add the new rects to the obstructions map
+    obstructions.erase(tech_layer); 
+    for (const auto& rect : rects) {
+      dbObstruction* new_obs = dbObstruction::create(block, 
```

Comment:
This seems to be a mistake and a local type should be used as these are just transient to the construction of the final OBS

Reply:
Yes, It was easy to started with the dbObstruction object just because it contain dbBox object and so I can use the writeBox without adding new write function.

I changed the values of the obstructions map to contain Rect objects and I modify the writeBox function to work with dbBox and Rect, I hoping it's ok in that way